File: typescript/src/routes/carts.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import app from '../index'
import { store } from '../stores'
import { loginTestUser, createAuthHeaders, TEST_USERS } from '../test-helpers/auth'
import type { components } from '../types/api'

type Cart = components['schemas']['Cart']
type ErrorResponse = { error: { code: string; message: string } }

describe('Carts API', () => {
  let authToken: string

  beforeEach(async () => {
    // Reset store to initial state before each test
    const storeInstance = new (store.constructor as any)()
    Object.setPrototypeOf(store, Object.getPrototypeOf(storeInstance))
    Object.assign(store, storeInstance)
    
    // Login to get auth token
    authToken = await loginTestUser(TEST_USERS.alice.email, TEST_USERS.alice.password)
  })

  describe('GET /carts/users/:userId', () => {
    it('should return cart for existing user', async () => {
      const res = await app.request('/carts/users/1', {
        headers: createAuthHeaders(authToken)
      })
      const json = await res.json() as Cart

      expect(res.status).toBe(200)
      expect(json).toHaveProperty('userId', '1')
      expect(json).toHaveProperty('items')
      expect(json.items).toHaveLength(0) // Initially empty
      expect(json).toHaveProperty('updatedAt')
    })

    it('should create empty cart for new user', async () => {
      const res = await app.request('/carts/users/999', {
        headers: createAuthHeaders(authToken)
      })
      const json = await res.json() as Cart

      expect(res.status).toBe(200)
      expect(json).toHaveProperty('userId', '999')
      expect(json.items).toHaveLength(0)
    })
  })

  describe('POST /carts/users/:userId/items', () => {
    it('should add item to cart', async () => {
      const addItemRequest = {
        productId: '1',
        quantity: 2
      }

      const res = await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify(addItemRequest),
      })
      const json = await res.json() as Cart

      expect(res.status).toBe(200)
      expect(json.items).toHaveLength(1)
      expect(json.items[0]).toEqual({
        productId: '1',
        quantity: 2
      })
    })

    it('should increase quantity when adding existing item', async () => {
      // First, add item
      await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify({ productId: '1', quantity: 2 }),
      })

      // Add same item again
      const res = await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify({ productId: '1', quantity: 3 }),
      })
      const json = await res.json() as Cart

      expect(res.status).toBe(200)
      expect(json.items).toHaveLength(1)
      expect(json.items[0].quantity).toBe(5) // 2 + 3
    })

    it('should return 404 for non-existent product', async () => {
      const addItemRequest = {
        productId: '999',
        quantity: 1
      }

      const res = await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify(addItemRequest),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(404)
      expect(json.error).toHaveProperty('code', 'NOT_FOUND')
    })

    it('should return 400 for insufficient stock', async () => {
      const addItemRequest = {
        productId: '1',
        quantity: 100 // More than available stock (10)
      }

      const res = await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify(addItemRequest),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(400)
      expect(json.error).toHaveProperty('code', 'INSUFFICIENT_STOCK')
    })
  })

  describe('PATCH /carts/users/:userId/items/:productId', () => {
    beforeEach(async () => {
      // Add item to cart first
      await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify({ productId: '1', quantity: 2 }),
      })
    })

    it('should update item quantity', async () => {
      const updateRequest = {
        quantity: 5
      }

      const res = await app.request('/carts/users/1/items/1', {
        method: 'PATCH',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify(updateRequest),
      })
      const json = await res.json() as Cart

      expect(res.status).toBe(200)
      expect(json.items[0].quantity).toBe(5)
    })

    it('should return 404 for item not in cart', async () => {
      const updateRequest = {
        quantity: 5
      }

      const res = await app.request('/carts/users/1/items/999', {
        method: 'PATCH',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify(updateRequest),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(404)
      expect(json.error).toHaveProperty('code', 'NOT_FOUND')
      expect(json.error.message).toContain('not found')
    })

    it('should return 400 for insufficient stock', async () => {
      const updateRequest = {
        quantity: 100 // More than available stock
      }

      const res = await app.request('/carts/users/1/items/1', {
        method: 'PATCH',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify(updateRequest),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(400)
      expect(json.error).toHaveProperty('code', 'INSUFFICIENT_STOCK')
    })
  })

  describe('DELETE /carts/users/:userId/items/:productId', () => {
    beforeEach(async () => {
      // Add items to cart first
      await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify({ productId: '1', quantity: 2 }),
      })
      await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify({ productId: '2', quantity: 1 }),
      })
    })

    it('should remove item from cart', async () => {
      const res = await app.request('/carts/users/1/items/1', {
        method: 'DELETE',
        headers: createAuthHeaders(authToken)
      })

      expect(res.status).toBe(204)

      // Verify item was removed
      const cartRes = await app.request('/carts/users/1', { headers: createAuthHeaders(authToken) })
      const cart = await cartRes.json() as Cart
      expect(cart.items).toHaveLength(1)
      expect(cart.items[0].productId).toBe('2')
    })

    it('should return 404 for item not in cart', async () => {
      const res = await app.request('/carts/users/1/items/999', {
        method: 'DELETE',
        headers: createAuthHeaders(authToken)
      })

      expect(res.status).toBe(404)
    })
  })

  describe('DELETE /carts/users/:userId/items', () => {
    beforeEach(async () => {
      // Add items to cart first
      await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify({ productId: '1', quantity: 2 }),
      })
      await app.request('/carts/users/1/items', {
        method: 'POST',
        headers: createAuthHeaders(authToken),
        body: JSON.stringify({ productId: '2', quantity: 1 }),
      })
    })

    it('should clear all items from cart', async () => {
      const res = await app.request('/carts/users/1/items', {
        method: 'DELETE',
        headers: createAuthHeaders(authToken)
      })

      expect(res.status).toBe(204)

      // Verify cart is empty
      const cartRes = await app.request('/carts/users/1', { headers: createAuthHeaders(authToken) })
      const cart = await cartRes.json() as Cart
      expect(cart.items).toHaveLength(0)
    })
  })

  describe('Authentication', () => {
    it('should return 401 for unauthenticated requests', async () => {
      const res = await app.request('/carts/users/1')
      
      expect(res.status).toBe(401)
      const json = await res.json() as ErrorResponse
      expect(json.error.code).toBe('UNAUTHORIZED')
    })

    it('should return 401 for invalid token', async () => {
      const res = await app.request('/carts/users/1', {
        headers: createAuthHeaders('invalid-token')
      })
      
      expect(res.status).toBe(401)
      const json = await res.json() as ErrorResponse
      expect(json.error.code).toBe('UNAUTHORIZED')
    })
  })
})

File: go/internal/handlers/carts_test.go
package handlers_test

import (
	"net/http"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestCartsService_Get(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should return cart for existing user", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "GET", "/carts/users/1", nil, token)
		assertStatus(t, rr, http.StatusOK)

		var cart map[string]any
		err := decodeJSON(rr, &cart)
		require.NoError(t, err)

		assert.Equal(t, "1", cart["userId"])
		assert.NotNil(t, cart["items"])
		items := cart["items"].([]any)
		assert.Len(t, items, 0) // Initially empty
		assert.NotEmpty(t, cart["id"])
		assert.NotEmpty(t, cart["createdAt"])
		assert.NotEmpty(t, cart["updatedAt"])
	})

	t.Run("should create empty cart for new user", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "GET", "/carts/users/999", nil, token)
		assertStatus(t, rr, http.StatusOK)

		var cart map[string]any
		err := decodeJSON(rr, &cart)
		require.NoError(t, err)

		assert.Equal(t, "999", cart["userId"])
		items := cart["items"].([]any)
		assert.Len(t, items, 0)
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		rr := makeRequest(t, server, "GET", "/carts/users/1", nil)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestCartsService_AddItem(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should add item to cart", func(t *testing.T) {
		addItem := map[string]any{
			"productId": "1",
			"quantity":  2,
		}

		rr := makeAuthenticatedRequest(t, server, "POST", "/carts/users/1/items", addItem, token)
		assertStatus(t, rr, http.StatusOK)

		var cart map[string]any
		err := decodeJSON(rr, &cart)
		require.NoError(t, err)

		items := cart["items"].([]any)
		assert.Len(t, items, 1)

		item := items[0].(map[string]any)
		assert.Equal(t, "1", item["productId"])
		assert.Equal(t, float64(2), item["quantity"])
	})

	t.Run("should increase quantity when adding existing item", func(t *testing.T) {
		// First, add item
		addItem := map[string]any{
			"productId": "1",
			"quantity":  2,
		}
		makeAuthenticatedRequest(t, server, "POST", "/carts/users/2/items", addItem, token)

		// Add same item again
		rr := makeAuthenticatedRequest(t, server, "POST", "/carts/users/2/items", addItem, token)
		assertStatus(t, rr, http.StatusOK)

		var cart map[string]any
		err := decodeJSON(rr, &cart)
		require.NoError(t, err)

		items := cart["items"].([]any)
		assert.Len(t, items, 1)

		item := items[0].(map[string]any)
		assert.Equal(t, float64(4), item["quantity"]) // 2 + 2
	})

	t.Run("should return 404 for non-existent product", func(t *testing.T) {
		addItem := map[string]any{
			"productId": "999",
			"quantity":  1,
		}

		rr := makeAuthenticatedRequest(t, server, "POST", "/carts/users/1/items", addItem, token)
		assertStatus(t, rr, http.StatusNotFound)
		assertErrorResponse(t, rr, "NOT_FOUND")
	})

	t.Run("should return 400 for insufficient stock", func(t *testing.T) {
		addItem := map[string]any{
			"productId": "1",
			"quantity":  100, // More than available stock (10)
		}

		rr := makeAuthenticatedRequest(t, server, "POST", "/carts/users/3/items", addItem, token)
		assertStatus(t, rr, http.StatusBadRequest)
		assertErrorResponse(t, rr, "INSUFFICIENT_STOCK")
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		addItem := map[string]any{
			"productId": "1",
			"quantity":  1,
		}

		rr := makeRequest(t, server, "POST", "/carts/users/1/items", addItem)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestCartsService_UpdateItem(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	// Setup: Add item to cart first
	setupCart := func(userID string) {
		addItem := map[string]any{
			"productId": "1",
			"quantity":  2,
		}
		makeAuthenticatedRequest(t, server, "POST", "/carts/users/"+userID+"/items", addItem, token)
	}

	t.Run("should update item quantity", func(t *testing.T) {
		setupCart("4")

		update := map[string]any{
			"quantity": 5,
		}

		rr := makeAuthenticatedRequest(t, server, "PATCH", "/carts/users/4/items/1", update, token)
		assertStatus(t, rr, http.StatusOK)

		var cart map[string]any
		err := decodeJSON(rr, &cart)
		require.NoError(t, err)

		items := cart["items"].([]any)
		item := items[0].(map[string]any)
		assert.Equal(t, float64(5), item["quantity"])
	})

	t.Run("should return 404 for item not in cart", func(t *testing.T) {
		update := map[string]any{
			"quantity": 5,
		}

		rr := makeAuthenticatedRequest(t, server, "PATCH", "/carts/users/5/items/999", update, token)
		assertStatus(t, rr, http.StatusNotFound)
		assertErrorResponse(t, rr, "NOT_FOUND")
	})

	t.Run("should return 400 for insufficient stock", func(t *testing.T) {
		setupCart("6")

		update := map[string]any{
			"quantity": 100,
		}

		rr := makeAuthenticatedRequest(t, server, "PATCH", "/carts/users/6/items/1", update, token)
		assertStatus(t, rr, http.StatusBadRequest)
		assertErrorResponse(t, rr, "INSUFFICIENT_STOCK")
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		update := map[string]any{
			"quantity": 5,
		}

		rr := makeRequest(t, server, "PATCH", "/carts/users/1/items/1", update)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestCartsService_RemoveItem(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	// Setup: Add items to cart
	setupCart := func(userID string) {
		// Add two different products
		makeAuthenticatedRequest(t, server, "POST", "/carts/users/"+userID+"/items", map[string]any{
			"productId": "1",
			"quantity":  2,
		}, token)
		makeAuthenticatedRequest(t, server, "POST", "/carts/users/"+userID+"/items", map[string]any{
			"productId": "2",
			"quantity":  1,
		}, token)
	}

	t.Run("should remove item from cart", func(t *testing.T) {
		setupCart("7")

		rr := makeAuthenticatedRequest(t, server, "DELETE", "/carts/users/7/items/1", nil, token)
		assertStatus(t, rr, http.StatusNoContent)

		// Verify item was removed
		cartRR := makeAuthenticatedRequest(t, server, "GET", "/carts/users/7", nil, token)
		var cart map[string]any
		err := decodeJSON(cartRR, &cart)
		require.NoError(t, err)

		items := cart["items"].([]any)
		assert.Len(t, items, 1)
		remainingItem := items[0].(map[string]any)
		assert.Equal(t, "2", remainingItem["productId"])
	})

	t.Run("should return 404 for item not in cart", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "DELETE", "/carts/users/8/items/999", nil, token)
		assertStatus(t, rr, http.StatusNotFound)
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		rr := makeRequest(t, server, "DELETE", "/carts/users/1/items/1", nil)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestCartsService_Clear(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	// Setup: Add items to cart
	setupCart := func(userID string) {
		makeAuthenticatedRequest(t, server, "POST", "/carts/users/"+userID+"/items", map[string]any{
			"productId": "1",
			"quantity":  2,
		}, token)
		makeAuthenticatedRequest(t, server, "POST", "/carts/users/"+userID+"/items", map[string]any{
			"productId": "2",
			"quantity":  1,
		}, token)
	}

	t.Run("should clear all items from cart", func(t *testing.T) {
		setupCart("9")

		rr := makeAuthenticatedRequest(t, server, "DELETE", "/carts/users/9/items", nil, token)
		assertStatus(t, rr, http.StatusNoContent)

		// Verify cart is empty
		cartRR := makeAuthenticatedRequest(t, server, "GET", "/carts/users/9", nil, token)
		var cart map[string]any
		err := decodeJSON(cartRR, &cart)
		require.NoError(t, err)

		items := cart["items"].([]any)
		assert.Len(t, items, 0)
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		rr := makeRequest(t, server, "DELETE", "/carts/users/1/items", nil)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestCartsService_Integration(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should handle complete cart workflow", func(t *testing.T) {
		userID := "100"

		// 1. Start with empty cart
		getEmptyCart := makeAuthenticatedRequest(t, server, "GET", "/carts/users/"+userID, nil, token)
		assertStatus(t, getEmptyCart, http.StatusOK)
		var cart map[string]any
		decodeJSON(getEmptyCart, &cart)
		items := cart["items"].([]any)
		assert.Len(t, items, 0)

		// 2. Add items
		addItem1 := makeAuthenticatedRequest(t, server, "POST", "/carts/users/"+userID+"/items", map[string]any{
			"productId": "1",
			"quantity":  2,
		}, token)
		assertStatus(t, addItem1, http.StatusOK)

		addItem2 := makeAuthenticatedRequest(t, server, "POST", "/carts/users/"+userID+"/items", map[string]any{
			"productId": "2",
			"quantity":  1,
		}, token)
		assertStatus(t, addItem2, http.StatusOK)

		// 3. Update quantity
		updateItem := makeAuthenticatedRequest(t, server, "PATCH", "/carts/users/"+userID+"/items/1", map[string]any{
			"quantity": 3,
		}, token)
		assertStatus(t, updateItem, http.StatusOK)

		// 4. Remove one item
		removeItem := makeAuthenticatedRequest(t, server, "DELETE", "/carts/users/"+userID+"/items/2", nil, token)
		assertStatus(t, removeItem, http.StatusNoContent)

		// 5. Verify final state
		getFinalCart := makeAuthenticatedRequest(t, server, "GET", "/carts/users/"+userID, nil, token)
		assertStatus(t, getFinalCart, http.StatusOK)
		decodeJSON(getFinalCart, &cart)
		finalItems := cart["items"].([]any)
		assert.Len(t, finalItems, 1)
		finalItem := finalItems[0].(map[string]any)
		assert.Equal(t, "1", finalItem["productId"])
		assert.Equal(t, float64(3), finalItem["quantity"])

		// 6. Clear cart
		clearCart := makeAuthenticatedRequest(t, server, "DELETE", "/carts/users/"+userID+"/items", nil, token)
		assertStatus(t, clearCart, http.StatusNoContent)

		// 7. Verify empty
		getCleared := makeAuthenticatedRequest(t, server, "GET", "/carts/users/"+userID, nil, token)
		assertStatus(t, getCleared, http.StatusOK)
		decodeJSON(getCleared, &cart)
		clearedItems := cart["items"].([]any)
		assert.Len(t, clearedItems, 0)
	})
}